staging/src/k8s.io/kubectl/pkg/cmd/auth/cani.go

```diff
@@ -325,6 +336,13 @@ func (o *CanIOptions) resourceFor(mapper meta.RESTMapper, resourceArg string) sc
 		}
 	}
 
+	if len(subresourceArg) > 0 && o.DiscoveryClient != nil {
```

Comment:
We may remove this null check, because we are failing above if discoveryClient gets error

Reply:
this is so the tests pass - I guess we could use a mock discovery there. Also I still need to work on the tests

Reply:
if we require a non-nil discovery client, then checking for nil here and taking different branches doesn't seem great

Reply:
I made sure to always supply the discovery client in the tests, so the null check is not necessary anymore.